From e9a9c263d7df663121cf3a4ea22206437eb89d38 Mon Sep 17 00:00:00 2001
From: Yelena <yelena@marvell.com>
Date: Sun, 6 Apr 2014 18:20:58 +0300
Subject: [PATCH 1575/1825] fix: neta: autoneg off/on ethtool command fixed

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit f56bc692b394227ec1173c18067da3032169cf24

	Fixed JIRA SYSTEMSW-435
	Enable/disable FlowControl autoneg trough
	general autoneg disable/enable command

Change-Id: Ifc61da093d7dbc91e01405f26ee4e1924f1f78fc
Signed-off-by: Yelena <yelena@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/6950
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .../mv_drivers_lsp/mv_neta/net_dev/mv_eth_tool.c   |   22 +++++++++++--------
 1 files changed, 13 insertions(+), 9 deletions(-)

diff --git a/arch/arm/plat-armada/mv_drivers_lsp/mv_neta/net_dev/mv_eth_tool.c b/arch/arm/plat-armada/mv_drivers_lsp/mv_neta/net_dev/mv_eth_tool.c
index 6c1e846..9e5b6ad 100644
--- a/arch/arm/plat-armada/mv_drivers_lsp/mv_neta/net_dev/mv_eth_tool.c
+++ b/arch/arm/plat-armada/mv_drivers_lsp/mv_neta/net_dev/mv_eth_tool.c
@@ -122,24 +122,28 @@ int mv_eth_tool_restore_settings(struct net_device *netdev)
 			err = mvEthPhyAdvertiseSet(phy_addr, priv->advertise_cfg);
 		/* Restart AN on PHY enables it */
 		if (!err) {
+			err = mvNetaFlowCtrlSet(priv->port, MV_ETH_FC_AN_SYM);
+			if (!err) {
+				err = mvEthPhyRestartAN(phy_addr, MV_ETH_TOOL_AN_TIMEOUT);
+				if (err == MV_TIMEOUT) {
+					MV_ETH_PORT_STATUS ps;
 
-			err = mvEthPhyRestartAN(phy_addr, MV_ETH_TOOL_AN_TIMEOUT);
-			if (err == MV_TIMEOUT) {
-				MV_ETH_PORT_STATUS ps;
+					mvNetaLinkStatus(priv->port, &ps);
 
-				mvNetaLinkStatus(priv->port, &ps);
-
-				if (!ps.linkup)
-					err = 0;
+					if (!ps.linkup)
+						err = 0;
+				}
 			}
 		}
 	} else if (priv->autoneg_cfg == AUTONEG_DISABLE) {
 		err = mvEthPhyDisableAN(phy_addr, phy_speed, phy_duplex);
 		if (!err)
+			err = mvNetaFlowCtrlSet(priv->port, MV_ETH_FC_ENABLE);
+		if (!err)
 			err = mvNetaSpeedDuplexSet(priv->port, mac_speed, mac_duplex);
-	} else {
+	} else
 		err = -EINVAL;
-	}
+
 	return err;
 }
 
-- 
1.7.5.4

